### PR TITLE
Fix: bar width when there is only one data point

### PR DESCRIPTION
**Pull request by souzolliveira**
## Description

- Adding condition to return a bar size when there is only one data point;
- Creating test for it.

## Related Issue

- Bug #3640 

## Motivation and Context

- I choose this bug to solve for a graduation issue and because it seems easy.

## How Has This Been Tested?

- I created a test to count one rectangle for one data point.

## Screenshots:

image

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] I have added a storybook story or extended an existing story to show my changes
- [x] All new and existing tests passed.

### CODE REVIEW

File: test/chart/BarChart.spec.tsx
Comment (ckifer) [inline test/chart/BarChart.spec.tsx:27]: can you confirm that this test doesn't pass before your change? If the rectangle gets rendered with a 0 width this would still pass

File: src/util/ChartUtils.ts
Comment (ckifer) [inline src/util/ChartUtils.ts:1269]: Instead of doing this maybe we re-think the type sent to this function. `BaseAxisProps` does not include `width` but `XAxis` and YAxis` props do. 

What if this type became a union of `XAxisProps` | `YAxisProps` | `ZAxisProps` and type narrowing can be done from there. That way we can add this with correct TS defs.

**Review by ckifer — CHANGES_REQUESTED** [review]
Thanks for contributing! A few questions

File: src/util/ChartUtils.ts
Comment (souzolliveira) [reply src/util/ChartUtils.ts]: Fixed

File: test/chart/BarChart.spec.tsx
Comment (souzolliveira) [reply test/chart/BarChart.spec.tsx:27]: Really, this condition can be satisfied in this case if the rectangle renders with a 0 width. I will verify.

Comment (souzolliveira) [reply test/chart/BarChart.spec.tsx:27]: The .recharts-rectangle isn't rendered when its width is zero, then, this test fails without my changes.

File: src/util/ChartUtils.ts
Comment (ckifer) [inline src/util/ChartUtils.ts:1255]: now the type isn't accurate because this function isn't called for only `XAxis`

We need a union type + type narrowing

So
```ts
// defined above or elsewhere
type CartesianAxisProps = XAxisProps | YAxisProps | ZAxisProps


// in function
 if (axis && 'width' in axis && axis.width && ticks) {
```

**Review by ckifer — CHANGES_REQUESTED** [review]

File: src/util/ChartUtils.ts
Comment (souzolliveira) [reply src/util/ChartUtils.ts]: I applied the fixes. Can you verify if it is correct now? Pls

File: src/util/ChartUtils.ts
Comment (ckifer) [inline src/util/ChartUtils.ts:1257]: the jsdoc belongs to the function not the type, please move the type above the jsdoc comment

**Review by ckifer — CHANGES_REQUESTED** [review]
looks good besides one tiny thing

File: src/util/ChartUtils.ts
Comment (ckifer) [inline src/util/ChartUtils.ts:39]: 😅 generally I'd put this below the re-export below but all good

**Review by ckifer — APPROVED** [review]
LGTM

This is an improvement but tbh numerical axis bar charts do not make much sense (unless the numbers are treated as categories). This is probably why this didn't work in the first place.

@nikolasrieble @PavelVanecek 
Any thoughts about this? If this produces an ugly chart (it does, see uncentered ticks) wouldn't it be better to throw an invariant when someone attempts to use two numerical axes for a bar chart? I see no reason not to ship this to improve current behavior, but maybe this is something we discourage in the next MV

**Comment by ckifer** [comment]
@souzolliveira thanks, going to wait for some other reviews

**Comment by ckifer** [comment]
also build is failing

File: src/util/ChartUtils.ts
Comment (souzolliveira) [reply src/util/ChartUtils.ts:39]: At the line 43?

**Comment by souzolliveira** [comment]
@ckifer I could increase the x-axis width to centralize the unique bar in the axis if you tell me the path to do this.
Did my changes break the tests that failed in the build?

**Comment by nikolasrieble** [comment]
> @ckifer I could increase the x-axis width to centralize the unique bar in the axis if you tell me the path to do this. Did my changes break the tests that failed in the build?

Yes, this specifically 

```
 <BarChart /> › Render a bar › renders a bar if size is limited
```

**Review by PavelVanecek — APPROVED** [review]
In my opinion, as a library we should allow use-cases that don't make sense to us. Maybe there's a valid chart that needs numerical axis and bars together, who knows.

**Comment by ckifer** [comment]
Agreed in principle, but the flexibility does add a lot of pressure on the side of the maintainers. 

We support this but is it expected to look a certain way or be functional? I guess it might be useful depending on the case and customizations applied. I wish the original creators had limited flexibility a bit (even though it's one of the things people like about the lib now) 😅

Doesn't matter for now though, just trying to think aloud a bit. Thanks for the input!

**Comment by ckifer** [comment]
@souzolliveira looks like test is still failing

**Comment by TheCleric** [comment]
To write in favor of having bar charts with numerical X and Y axes, here are some ones I am using for a project:

https://bferentz.vercel.app/

In a lot of charts on that page, I am comparing a numeric value (potential points scored in a game) against another numeric value (the % chance of those points being score).

**Comment by ckifer** [comment]
@TheCleric nice, thanks for the data point (and cool site!)

At small scale things look relatively non-functional but these types of examples help challenge (at least my) perception of what people are able to do with the lib. Kudos for creativity and we'll make sure number support doesn't go away

**Comment by ckifer** [comment]
@souzolliveira any chance you could merge the latest and fix this test?

**Comment by souzolliveira** [comment]
> @souzolliveira any chance you could merge the latest and fix this test?

Yes, sorry for the delay, I will update my branch and correct the tests.

**Comment by ckifer** [comment]
No worries at all

File: test/chart/BarChart.spec.tsx
Comment (nikolasrieble) [inline test/chart/BarChart.spec.tsx:22]:
```suggestion
```

File: src/util/ChartUtils.ts
Comment (nikolasrieble) [inline src/util/ChartUtils.ts:1299]:
```suggestion
      const orderedTicks = sortBy(ticks, o => o.coordinate);
```

**Comment by nikolasrieble** [comment]
Trying to land this, I merged master via UI. Testing this locally, tests are failing 🤔 
Strangely the CI was not triggered by actions in the UI.

**Review by nikolasrieble — CHANGES_REQUESTED** [review]
Lets make sure to have all tests pass before this lands.

**Comment by ckifer** [comment]
Going to close this as stale for now, potentially @graup will take it over

**Pull request closed without merging**
